test/test-observers.js

```diff
@@ -120,4 +120,43 @@ exports["test weak observers are GC-ed on unload"] = (assert, end) => {
   });
 };
 
-require("sdk/test").run(exports);
\ No newline at end of file
+exports["test weak observer unsubscribe"] = (assert, end) => {
+  const loader = Loader(module);
+  const { Observer, observe, subscribe, unsubscribe } = loader.require("sdk/core/observer");
+  const { WeakReference } = loader.require("sdk/core/reference");
+
+  let sawNotification = false;
+
+  const WeakObserver = Class({
+    extends: Observer,
+    implements: [WeakReference],
+    observe: function() {
+      sawNotification = true;
+    }
+  });
+  observe.define(WeakObserver, (x, ...rest) => x.observe(...rest));
+
+  let observer = new WeakObserver;
+  subscribe(observer, "test-topic");
+
+  notifyObservers(null, "test-topic", null);
+  assert.ok(sawNotification, "Should have seen notification before GC");
+  sawNotification = false;
+
+  loader.require("sdk/test/memory").gc();
+
+  notifyObservers(null, "test-topic", null);
```

Comment:
I don't understand what that adds to the test. The observer service is
synchronous, it guarantees that all observers will have been called when
notifyObservers returns, adding a second observer seems like a no-no.
On Feb 10, 2015 11:12 PM, "Erik Vold" <notifications@github.com> wrote:

> In test/test-observers.js
> <https://github.com/mozilla/addon-sdk/pull/1700#discussion_r24477411>:
>
> > +    observe: function() {
> > +      sawNotification = true;
> > +    }
> > +  });
> > +  observe.define(WeakObserver, (x, ...rest) => x.observe(...rest));
> > +
> > +  let observer = new WeakObserver;
> > +  subscribe(observer, "test-topic");
> > +
> > +  notifyObservers(null, "test-topic", null);
> > +  assert.ok(sawNotification, "Should have seen notification before GC");
> > +  sawNotification = false;
> > +
> > +  loader.require("sdk/test/memory").gc();
> > +
> > +  notifyObservers(null, "test-topic", null);
>
> we should yield on promise from another observer here either for the same
> topic or another one, to keep the test async.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mozilla/addon-sdk/pull/1700/files#r24477411>.
>